### PR TITLE
#1612 Separating Menas UI and API

### LINKED ISSUE #1612: Separate Menas into two Tomcat deployable applications

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #1620 
Closes #1649 
Separating menas into a client and API part.

Menas was separated into an API and UI part. The API will just be the Spring application serving REST endpoints and the UI will just serve static files to the browser and all other requests will be done to the API. Using cookies was moved to the UI, making the communication stateless. The API just returns the JWT and X-CSRF-TOKEN headers and the cookie is being set in JavaScript in the browser.

The UI is now built with OpenUI5 tools, but building the UI part is done by Maven and it generates two things: the `dist` path for static files which are supposed to be passed to the nginx container or any other CDN and the war in `target` which contains the same files found in `dist`.

After running  `mvn clean install`, running it locally can be done in two ways:
1. Use the generated source files in `menas-web/ui/dist` to pass them to the Docker nginx container and deploy the api application in a container too.  The nginx container will need the following runtime environment variables:

- `API_URL` , mandatory, needed for knowing the API host

-  `PRIVATE_KEY`, `CERTIFICATE`, `CA_CHAIN` and `DNS_NAME` which are optional and are used in case you want to use TLS certificates 

2. Deploy the generated war files to Tomcat, each having its path. It can be advised to either modify the `"apiUrl": "${API_URL}"` line in package.json in the UI client with the correct host for the api or modify it in the generated war.
3. A combination between the two if wanted

### CODE REVIEW

File: menas/Dockerfile
Comment (dk1844) [inline menas/Dockerfile:28]: This should probably be removed before merging.

**Review by dk1844 — APPROVED** [review]
(just read through)

File: menas/Dockerfile
Comment (AdrianOlosutean) [reply menas/Dockerfile:28]: I think it depends. I thought it could help for troubleshooting locally

File: menas-web/ui/components/login/loginDetail.controller.js
Comment (benedeki) [reply menas-web/ui/components/login/loginDetail.controller.js:152]: Some normalization and checking could be done on setting the property 🤔

File: menas/Dockerfile
Comment (dk1844) [reply menas/Dockerfile:28]: Makes sense.

File: menas-web/ui/components/login/loginDetail.controller.js
Comment (benedeki) [inline menas-web/ui/components/login/loginDetail.controller.js:152]: If it's called `apiUrl` why to add the subfolder `api` everywhere?

File: menas-web/ui/index.html
Comment (benedeki) [inline menas-web/ui/index.html:60]: **Proposition:** I wouldn't bother with Lineage/Spline now. Will see how it be integrated.

File: menas-web/Dockerfile
Comment (Zejnilovic) [inline menas-web/Dockerfile:9]: You know when you write something just so you don't forget and then it propagates everywhere. That is the history of this comment. Probably shouldn't be here but in some documentation page.

File: menas-web/Dockerfile
Comment (Zejnilovic) [inline menas-web/Dockerfile:19]: No secure connection?

File: menas-web/pom.xml
Comment (Zejnilovic) [inline menas-web/pom.xml:67]: Should this be left here?

File: menas-web/pom.xml
Comment (Zejnilovic) [inline menas-web/pom.xml:53]: Can `id`'s have spaces in them?

File: menas-web/ui/theme/sap/tnt/themes/sap_belize/library-parameters.json
Comment (Zejnilovic) [inline menas-web/ui/theme/sap/tnt/themes/sap_belize/library-parameters.json:1]: where does this come from? Could it be formatted better?

File: menas/Dockerfile
Comment (Zejnilovic) [reply menas/Dockerfile:28]: Actually, the comment can be removed and left as is (`EXPOSE 5005`). When you then start the image you choose if to bind it or not

**Review by Zejnilovic — COMMENTED** [review]
Had just a quick look for now

File: menas-web/ui/components/login/loginDetail.controller.js
Comment (AdrianOlosutean) [reply menas-web/ui/components/login/loginDetail.controller.js:152]: It makes sense. Also, does it make sense to take apiUrl from package.json or it may be better to have another file like config.json?

File: menas-web/ui/theme/sap/tnt/themes/sap_belize/library-parameters.json
Comment (AdrianOlosutean) [reply menas-web/ui/theme/sap/tnt/themes/sap_belize/library-parameters.json:1]: So the story is that I could not point to the npm dependency for the theme: `"@openui5/themelib_sap_belize": "^1.67.1"`because it lacks the library.css and library-parameters.json files which are build somehow but I could not figure out how, so I added the needed files to the source code. This can be considered a first step, but it's an open discussion how to properly provide these files.

File: menas-web/pom.xml
Comment (AdrianOlosutean) [reply menas-web/pom.xml:53]: Yes

File: menas-web/pom.xml
Comment (AdrianOlosutean) [reply menas-web/pom.xml:67]: I may need to figure out how to properly sync the maven version with the package.json version. Should this be at the build time?

File: menas-web/Dockerfile
Comment (AdrianOlosutean) [reply menas-web/Dockerfile:19]: Probably makes sense to be secure. Another question would be: would it make sense to use an Nginx server to serve these static files or it would be better to properly test the generated war?

File: menas-web/ui/components/login/loginDetail.controller.js
Comment (AdrianOlosutean) [reply menas-web/ui/components/login/loginDetail.controller.js:152]: Good point. As an idea, should Menas also keep the `/api` paths?

File: menas-web/ui/service/GenericService.js
Comment (AdrianOlosutean) [reply menas-web/ui/service/GenericService.js:33]: What could be preferable, a more readable code or an explanation comment?

File: menas/src/main/resources/application.properties.template
Comment (Zejnilovic) [inline menas/src/main/resources/application.properties.template:78]: Since you removed this. I cannot build the project as it is a mandatory property for Run Service

File: menas-web/ui/index.html
Comment (Zejnilovic) [reply menas-web/ui/index.html:52]: If Spline is not working we cannot merge this. Spline on-prem still needs to work

File: menas-web/start_menas_web.sh
Comment (Zejnilovic) [inline menas-web/start_menas_web.sh:12]: missing new line

File: menas-web/ui/package.json
Comment (Zejnilovic) [inline menas-web/ui/package.json:12]: This should not be here. You should not have to care that you made changes to git each time you want to test deploy. There should be another config file or something.

File: menas-web/ui/package.json
Comment (Zejnilovic) [inline menas-web/ui/package.json:3]: Cannot this be taken or updated automatically? I don't see us updating this each time we release new version.

**Review by Zejnilovic — COMMENTED** [review]
Could you please add more description to the PR, like documentation how to run?

File: menas-web/ui/package.json
Comment (AdrianOlosutean) [reply menas-web/ui/package.json:3]: There was at a certain point a sync tool, I should try it again

File: menas-web/ui/generic/functions.js
Comment (Zejnilovic) [inline menas-web/ui/generic/functions.js:35]: This prints null

File: menas-web/ui/service/OozieService.js
Comment (lokm01) [inline menas-web/ui/service/OozieService.js:21]: Shouldn't this also be prefixed with apiUrl?

**Review by lokm01 — COMMENTED** [review]
Checked the code **very briefly**.

Agree with the other comments.

File: menas-web/ui/service/OozieService.js
Comment (AdrianOlosutean) [reply menas-web/ui/service/OozieService.js:21]: It's prefixed [here](https://github.com/AbsaOSS/enceladus/pull/1620/files#diff-0818e6ba5065d9f79bee913ebd671ea33847d6e51943e44385f68e74310fb2b5R51) only once

File: menas-web/ui/package.json
Comment (Zejnilovic) [reply menas-web/ui/package.json:12]: API_URL for a system variable is too broad of a name. Please rename to `MENAS_API_URL`

File: menas-web/ui/index.html
Comment (Zejnilovic) [inline menas-web/ui/index.html:51]: URLs should not have trailing `/`

```suggestion
            window.apiUrl = json.config.apiUrl + "/api";
```

This is a general standard to make sure that everyone can contact paths in URL without thinking about it.

File: menas-web/ui/components/schema/schemaDetail.controller.js
Comment (Zejnilovic) [inline menas-web/ui/components/schema/schemaDetail.controller.js:391]: previous URLs start with / this one doesn't. Please align the URLs

File: menas-web/ui/package.json
Comment (AdrianOlosutean) [reply menas-web/ui/package.json:12]: It's being set in the `start_menas_web.sh` script by:
`envsubst < /usr/share/nginx/html/package.json > /usr/share/nginx/html/package-new.json
rm /usr/share/nginx/html/package.json
mv /usr/share/nginx/html/package-new.json /usr/share/nginx/html/package.json`.

The script is run in the Dockerfile.

File: menas-web/ui/index.html
Comment (AdrianOlosutean) [reply menas-web/ui/index.html:51]: I was more on the thinking to add the trailing only once in /api/, so that anywhere else it wouldn't be needed

File: menas-web/ui/package.json
Comment (Zejnilovic) [reply menas-web/ui/package.json:12]: So you need to build `menas_web.war` or docker image per environment?

File: menas-web/ui/index.html
Comment (Zejnilovic) [reply menas-web/ui/index.html:51]: - `/api/` is a 2 segment URI, with api and empty as the segments. This immediately stinks
- When building URIs (concatenating them), look at spring, you are expected to start with a `/` but not have any at the end. This in my experience is a rule followed by people regardless of platform/language

Please remove the trailing / from your URIs.

File: menas-web/ui/components/schema/schemaDetail.controller.js
Comment (AdrianOlosutean) [reply menas-web/ui/components/schema/schemaDetail.controller.js:391]: I thought it was comitted

File: menas-web/ui/package.json
Comment (AdrianOlosutean) [reply menas-web/ui/package.json:12]: No, the docker image just needs an environment variable `API_URL` which will be filled into package.json. When it comes to the war type deployment, the war can also be automatically modified to substitute the `API_URL`

File: menas-web/pom.xml
Comment (Zejnilovic) [inline menas-web/pom.xml:18]: Could we add `dockerfile-maven-plugin` here?

Comment (Zejnilovic) [reply menas-web/pom.xml:18]: With buildArgs of SPLINE and API_URL

Comment (AdrianOlosutean) [reply menas-web/pom.xml:18]: Good idea, I will add it here too

Comment (Zejnilovic) [reply menas-web/pom.xml:18]: Had a call. I misunderstood what is happening. The docker plugin will be all we need.

**Review by dk1844 — APPROVED** [review]
(read the code)

**Review by dk1844 — APPROVED** [review]

**Pull request merged**
